app/src/main/java/net/gini/android/gvlexample/gini/SingleDocumentAnalyzer.java

```diff
@@ -0,0 +1,257 @@
+    }
+
+    /**
+     * <p>
+     *     Analyzes a new document only, if there was no previous analysis or the previous one was cancelled.
+     * </p>
+     */
+    public void analyzeDocument(Document document, final DocumentAnalysisListener listener) {
+        LOG.debug("Start analyzing document");
+        if (mAnalyzer != null) {
+            if (!mAnalyzer.isCancelled()) {
+                LOG.debug("Analysis in progress, only changing the listener");
+                setListener(listener);
+                return;
+            }
+            mAnalyzer.setListener(null);
+        }
+        LOG.debug("Starting a new analysis");
+
+        mAnalyzer = new Analyzer(mGiniApi.getDocumentTaskManager());
+        setListener(listener);
+        mAnalyzer.analyze(document);
+    }
+
+    public void cancelAnalysis() {
+        LOG.debug("Canceling analysis");
+        if (mAnalyzer != null) {
+            mAnalyzer.cancel();
+            LOG.debug("Analysis cancelled");
+        } else {
+            LOG.debug("No running analysis to cancel");
+        }
+    }
+
+    public void removeListener() {
+        LOG.debug("Removing listener");
+        if (mAnalyzer != null) {
+            mAnalyzer.setListener(null);
+            LOG.debug("Listener removed");
+        } else {
+            LOG.debug("No running analysis to remove listener from");
+        }
+    }
+
+    public void setListener(final DocumentAnalysisListener listener) {
+        LOG.debug("Setting listener");
+        if (mAnalyzer != null) {
+            mAnalyzer.setListener(new DocumentAnalysisListener() {
+                @Override
+                public void onExtractionsReceived(Map<String, SpecificExtraction> extractions) {
+                    logExtractions(extractions);
+                    listener.onExtractionsReceived(extractions);
+                }
+
+                @Override
+                public void onException(Exception exception) {
+                    LOG.error("Analysis failed", exception);
+                    listener.onException(exception);
+                }
+            });
+            LOG.debug("Listener set");
+        } else {
+            LOG.debug("No running analysis to set listener on");
+        }
+    }
+
+    public net.gini.android.models.Document getGiniApiDocument() {
+        LOG.debug("Getting Gini Api Document");
+        if (mAnalyzer != null){
+            if(mAnalyzer.isCompleted()) {
+                if (mAnalyzer.getGiniApiDocument() == null) {
+                    LOG.debug("No Gini Api Document Available: analysis failed");
+                }
+                return mAnalyzer.getGiniApiDocument();
+            } else {
+                LOG.debug("No Gini Api Document Available: analysis in progress");
+            }
+        } else {
+            LOG.debug("No Gini Api Document Available: no analysis started");
+        }
+        return null;
+    }
+
+    public interface DocumentAnalysisListener {
```

Comment:
Interfaces use to be on the top of the class in Java, right?

Reply:
There's no rule for that, I use the default order in IntelliJ, which puts inner interfaces and classes at the end, which is fine for me since I can say, that first I want to see the fields and methods of the class to know what it does and the details I can check at the end.